css/css-view-transitions/iframe-and-main-frame-transition-old-main* fails.

mattwoodrow opened the pull request:
#### ee9b30c26a2450890015f759bfbb7cf7a3c739c4
<pre>
css/css-view-transitions/iframe-and-main-frame-transition-old-main* fails.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293275">https://bugs.webkit.org/show_bug.cgi?id=293275</a>
&lt;<a href="https://rdar.apple.com/151667622">rdar://151667622</a>&gt;

Reviewed by Tim Nguyen.

Store the computed PaintBehavior for VT snapshotting on the view (as well as
providing it to the layer), so that it gets propagated to subdocuments.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:

Canonical link: <a href="https://commits.webkit.org/295251@main">https://commits.webkit.org/295251@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f188a5b3241b5af923f381f9faf3327d6c75a65

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104339 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24043 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14457 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109540 "Hash 0f188a5b for PR 45628 does not build (failure)") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55007 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106379 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32590 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/109540 "Hash 0f188a5b for PR 45628 does not build (failure)") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107345 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18975 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94159 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/109540 "Hash 0f188a5b for PR 45628 does not build (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18769 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12232 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54367 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88474 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12287 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111922 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31496 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23204 "Found 1 new test failure: media/no-autoplay-with-user-gesture-requirement.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88264 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31860 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90403 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87928 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22434 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32821 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10588 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25964 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31424 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36749 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31218 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34554 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32778 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderWidget.cpp:
```diff
@@ -257,12 +257,15 @@ void RenderWidget::paintContents(PaintInfo& paintInfo, const LayoutPoint& paintO
+    std::optional<OptionSet<PaintBehavior>> oldBehavior = PaintBehavior::Normal;
+    if (RefPtr frameView = dynamicDowncast<LocalFrameView>(m_widget)) {
+        oldBehavior = frameView->paintBehavior();
+        frameView->setPaintBehavior(*oldBehavior | (paintInfo.paintBehavior & flagsToCopy));
```
maybe name it `oldFrameBehavior` for slightly more clarity, since there's multiple behaviors involved here

nt1m commented on Source/WebCore/rendering/RenderWidget.cpp:
```diff
@@ -257,12 +257,15 @@ void RenderWidget::paintContents(PaintInfo& paintInfo, const LayoutPoint& paintO
+        PaintBehavior::FlattenCompositingLayers,
+        PaintBehavior::Snapshotting
+    };
+    std::optional<OptionSet<PaintBehavior>> oldBehavior = PaintBehavior::Normal;
```
I'm not sure I understand how the `std:optional` is useful here. Would it ever be `std::nullopt` if you set it to `PaintBehavior::Normal` ?

mattwoodrow replied:
Good catch, this was a bug.

I ended up finding a better way to write this entire PR though, so no longer relevant :)

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -439,19 +439,26 @@ static RefPtr<ImageBuffer> snapshotElementVisualOverflowClippedToViewport(LocalF
+    auto oldPaintBehavior = frameView->paintBehavior();
+    frameView->setPaintBehavior(oldPaintBehavior | PaintBehavior::FlattenCompositingLayers | PaintBehavior::Snapshotting);
+
     auto paintFlags = RenderLayer::paintLayerPaintingCompositingAllPhasesFlags();
     paintFlags.add(RenderLayer::PaintLayerFlag::TemporaryClipRects);
     paintFlags.add(RenderLayer::PaintLayerFlag::AppliedTransform);
     paintFlags.add(RenderLayer::PaintLayerFlag::PaintingSkipDescendantViewTransition);
-    layerRenderer->layer()->paint(buffer->context(), paintRect, LayoutSize(), { PaintBehavior::FlattenCompositingLayers, PaintBehavior::Snapshotting }, nullptr, paintFlags);
+    layerRenderer->layer()->paint(buffer->context(), paintRect, LayoutSize(), frameView->paintBehavior(), nullptr, paintFlags);
```
Is there a side effect to taking the old paint behaviors from the frameView? (maybe not, just wondering since we didn't do it before)

mattwoodrow replied:
Shouldn't be! The frameView's value should be the same as what gets passed to the layer (some consumers look it up that way, instead of having it passed down the stack), and that's what we had wrong.

nt1m reviewed (approved):
Does it need work on site isolation? Probably would be good to inform relevant people.

mattwoodrow committed:
css/css-view-transitions/iframe-and-main-frame-transition-old-main* fails.
https://bugs.webkit.org/show_bug.cgi?id=293275
<rdar://151667622>

Reviewed by Tim Nguyen.

Store the computed PaintBehavior for VT snapshotting on the view (as well as
providing it to the layer), so that it gets propagated to subdocuments.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:

Canonical link: https://commits.webkit.org/295251@main
